File: app/src/main/python/ai_service.py
import json
import re
import sys
import threading
import urllib.request
from settings import active_settings

# Pre-compiled script detection regular expressions for speed
RE_DEVANAGARI = re.compile(r"[\u0900-\u097F]")
RE_TAMIL = re.compile(r"[\u0B80-\u0BFF]")
RE_TELUGU = re.compile(r"[\u0C00-\u0C7F]")
RE_BENGALI = re.compile(r"[\u0980-\u09FF]")


class GeminiAIService:

    GEMINI_API_URL = "https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent"

    def __init__(self):
        self._lock = threading.Lock()
        self._is_summarizing = False

    def get_api_key(self):
        return getattr(active_settings, "GEMINI_API_KEY", "").strip()

    def detect_script_language(self, text):
        """Ultra-fast compiled script language detection."""
        if not text:
            return "en"

        if RE_DEVANAGARI.search(text):
            return "hi"
        elif RE_TAMIL.search(text):
            return "ta"
        elif RE_TELUGU.search(text):
            return "te"
        elif RE_BENGALI.search(text):
            return "bn"

        return "en"

    def make_gemini_request(self, prompt, base64_image=None):
        """Helper to send a JSON payload to Google AI Studio Gemini API using urllib with strict timeout handling."""
        api_key = self.get_api_key()
        if not api_key:
            return "Google AI Studio API Key not set. Please set your API key in settings."

        url = f"{self.GEMINI_API_URL}?key={api_key}"

        parts = []
        if base64_image:
            parts.append({
                "inline_data": {
                    "mime_type": "image/jpeg",
                    "data": base64_image
                }
            })

        parts.append({"text": prompt})

        payload = {
            "contents": [{
                "parts": parts
            }]
        }

        try:
            data_bytes = json.dumps(payload).encode("utf-8")
            req = urllib.request.Request(
                url,
                data=data_bytes,
                headers={"Content-Type": "application/json"},
                method="POST"
            )

            with urllib.request.urlopen(req, timeout=5) as response:
                res_body = response.read().decode("utf-8")
                res_json = json.loads(res_body)

                candidates = res_json.get("candidates", [])
                if candidates:
                    parts = candidates[0].get("content", {}).get("parts", [])
                    if parts:
                        return parts[0].get("text", "").strip()

                return "AI response empty."

        except Exception as e:
            print(f"Error calling Gemini API: {e}", file=sys.stderr)
            return "AI Service unavailable."

    def summarize_screen_async(self, screen_text, callback, error_callback=None):
        """Asynchronously summarizes screen text without blocking UI or Accessibility loops."""
        with self._lock:
            if self._is_summarizing:
                if error_callback:
                    error_callback("Already summarizing. Please wait.")
                return
            self._is_summarizing = True

        def run():
            try:
                summary = self.summarize_screen(screen_text)
                callback(summary)
            except Exception as e:
                if error_callback:
                    error_callback(f"Error: {str(e)}")
            finally:
                with self._lock:
                    self._is_summarizing = False

        threading.Thread(target=run, daemon=True).start()

    def summarize_screen(self, screen_text):
        if not screen_text or not screen_text.strip():
            return "Screen has no readable text to summarize."

        prompt = (
            "You are an AI assistant for a blind screen reader user. "
            "Summarize the following mobile app screen content into 1 or 2 concise, clear sentences:\n\n"
            f"{screen_text}"
        )
        return self.make_gemini_request(prompt)

    def translate_text(self, text, target_language="Hindi"):
        if not text or not text.strip():
            return text

        prompt = (
            f"Translate the following text accurately into {target_language}. "
            "Output ONLY the translated string without commentary:\n\n"
            f"{text}"
        )
        return self.make_gemini_request(prompt)

    def rewrite_simplified(self, text):
        if not text or not text.strip():
            return text

        prompt = (
            "You are an accessibility assistant for visually impaired users. "
            "Simplify and rewrite the following text so it is very easy to understand:\n\n"
            f"{text}"
        )
        return self.make_gemini_request(prompt)

    def describe_unlabeled_icon(self, class_name, context_text=""):
        prompt = (
            "An unlabeled UI button was tapped in a mobile app. "
            f"Widget type: '{class_name}'. Nearby context: '{context_text}'. "
            "In 3 to 5 words, describe what this button likely does."
        )
        return self.make_gemini_request(prompt)

    def describe_image_b64(self, base64_jpeg):
        prompt = (
            "Describe what is in this image concisely in 1 or 2 sentences for a blind user."
        )
        return self.make_gemini_request(prompt, base64_image=base64_jpeg)


# Global AI Service instance
ai_service_instance = GeminiAIService()

File: app/src/main/java/com/indian/screenreader/core/AiService.kt
package com.indian.screenreader.core

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL
import java.security.MessageDigest
import java.util.concurrent.Executors
import java.util.concurrent.atomic.AtomicBoolean

object AiService {
    private const val TAG = "AiService"
    private const val GEMINI_API_URL = "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent"
    
    // Separate executors so translation never blocks summarize/describe and vice versa
    private val aiExecutor = Executors.newFixedThreadPool(2)
    private val translateExecutor = Executors.newSingleThreadExecutor()
    private val isSummarizing = AtomicBoolean(false)
    private val isDescribingImage = AtomicBoolean(false)
    // Tracks the utterance ID of the last translation so stale results can be dropped
    @Volatile private var latestTranslationId = 0L

    private fun getApiKey(): String {
        return Settings.GEMINI_API_KEY.trim()
    }

    private fun makeGeminiRequest(prompt: String, base64Image: String? = null): String {
        val apiKey = getApiKey()
        if (apiKey.isEmpty()) {
            return "Google AI Studio API Key not set. Please set your API key in settings."
        }

        val urlString = "$GEMINI_API_URL?key=$apiKey"
        
        try {
            val url = URL(urlString)
            val connection = url.openConnection() as HttpURLConnection
            connection.requestMethod = "POST"
            connection.setRequestProperty("Content-Type", "application/json")
            connection.doOutput = true
            connection.connectTimeout = 20000
            connection.readTimeout = 20000

            val partsArray = JSONArray()
            
            if (base64Image != null) {
                val inlineData = JSONObject().apply {
                    put("mime_type", "image/jpeg")
                    put("data", base64Image)
                }
                val part = JSONObject().apply {
                    put("inline_data", inlineData)
                }
                partsArray.put(part)
            }

            val textPart = JSONObject().apply {
                put("text", prompt)
            }
            partsArray.put(textPart)

            val content = JSONObject().apply {
                put("parts", partsArray)
            }
            val contentsArray = JSONArray().apply {
                put(content)
            }
            val payload = JSONObject().apply {
                put("contents", contentsArray)
            }

            OutputStreamWriter(connection.outputStream).use { writer ->
                writer.write(payload.toString())
            }

            val responseCode = connection.responseCode
            if (responseCode == HttpURLConnection.HTTP_OK) {
                val responseBody = connection.inputStream.bufferedReader().use { it.readText() }
                val responseJson = JSONObject(responseBody)
                val candidates = responseJson.optJSONArray("candidates")
                if (candidates != null && candidates.length() > 0) {
                    val candidate = candidates.getJSONObject(0)
                    val contentObj = candidate.optJSONObject("content")
                    val resParts = contentObj?.optJSONArray("parts")
                    if (resParts != null && resParts.length() > 0) {
                        return resParts.getJSONObject(0).optString("text", "").trim()
                    }
                }
                return "AI response empty."
            } else {
                val errorBody = connection.errorStream?.bufferedReader()?.use { it.readText() }
                Log.e(TAG, "Gemini API Error: $responseCode - $errorBody")
                return "AI Service unavailable. Status code $responseCode."
            }

        } catch (e: Exception) {
            Log.e(TAG, "Error calling Gemini API", e)
            return "AI Service unavailable."
        }
    }

    private val summaryCache = android.util.LruCache<String, String>(20)

    fun summarizeScreenAsync(screenText: String, callback: (String) -> Unit, errorCallback: (String) -> Unit) {
        if (screenText.isBlank()) {
            callback("Screen has no readable text to summarize.")
            return
        }

        // Check response cache using SHA-256 hash to prevent collisions
        val digest = MessageDigest.getInstance("SHA-256")
        val hashBytes = digest.digest(screenText.toByteArray())
        val cacheKey = hashBytes.joinToString("") { "%02x".format(it) }
        val cached = summaryCache.get(cacheKey)
        if (cached != null) {
            callback("(Cached) $cached")
            return
        }

        if (!isSummarizing.compareAndSet(false, true)) {
            errorCallback("Already summarizing. Please wait.")
            return
        }

        aiExecutor.execute {
            try {
                if (getApiKey().isBlank()) {
                    callback("AI Offline (API Key not set). Raw screen text: " + screenText.take(200))
                    return@execute
                }
                val summary = summarizeScreen(screenText)
                if (summary.isNotBlank() && !summary.startsWith("AI Service")) {
                    summaryCache.put(cacheKey, summary)
                }
                callback(summary)
            } catch (e: Exception) {
                callback("AI Offline. Screen text: " + screenText.take(200))
            } finally {
                isSummarizing.set(false)
            }
        }
    }

    private fun summarizeScreen(screenText: String): String {
        val prompt = "You are an AI assistant for a blind screen reader user. " +
                "Summarize the following mobile app screen content into 1 or 2 concise, clear sentences:\n\n" +
                screenText
        return makeGeminiRequest(prompt)
    }

    fun translateTextAsync(text: String, targetLanguage: String = "Hindi", callback: (String) -> Unit) {
        if (text.isBlank()) {
            callback(text)
            return
        }
        val myId = System.currentTimeMillis()
        latestTranslationId = myId

        translateExecutor.execute {
            try {
                if (getApiKey().isBlank()) {
                    if (latestTranslationId == myId) callback(text)
                    return@execute
                }
                val prompt = "Translate the following text accurately into $targetLanguage. " +
                        "Output ONLY the translated string without commentary:\n\n$text"
                val result = makeGeminiRequest(prompt)
                if (latestTranslationId == myId) {
                    callback(result)
                }
            } catch (e: Exception) {
                if (latestTranslationId == myId) callback(text) // fallback to original
            }
        }
    }

    fun rewriteSimplifiedAsync(text: String, callback: (String) -> Unit, errorCallback: (String) -> Unit) {
        if (text.isBlank()) {
            callback(text)
            return
        }
        aiExecutor.execute {
            try {
                if (getApiKey().isBlank()) {
                    callback(text)
                    return@execute
                }
                val prompt = "You are an accessibility assistant for visually impaired users. " +
                        "Simplify and rewrite the following text so it is very easy to understand:\n\n$text"
                callback(makeGeminiRequest(prompt))
            } catch (e: Exception) {
                callback(text)
            }
        }
    }

    fun describeImageB64Async(base64Jpeg: String, callback: (String) -> Unit, errorCallback: (String) -> Unit) {
        if (!isDescribingImage.compareAndSet(false, true)) {
            errorCallback("Already describing an image. Please wait.")
            return
        }
        aiExecutor.execute {
            try {
                if (getApiKey().isBlank()) {
                    callback("AI Offline. Please set your Gemini API Key in settings to describe images.")
                    return@execute
                }
                val prompt = "Describe what is in this image concisely in 1 or 2 sentences for a blind user."
                callback(makeGeminiRequest(prompt, base64Jpeg))
            } catch (e: Exception) {
                errorCallback(e.message ?: "Unknown error")
            } finally {
                isDescribingImage.set(false)
            }
        }
    }

    fun extractImageTextB64Async(base64Jpeg: String, callback: (String) -> Unit, errorCallback: (String) -> Unit) {
        aiExecutor.execute {
            try {
                if (getApiKey().isBlank()) {
                    callback("AI OCR Offline. Please set your Gemini API Key in settings.")
                    return@execute
                }
                val prompt = "Extract and transcribe all readable text from this image accurately. Output ONLY the extracted text:"
                val result = makeGeminiRequest(prompt, base64Jpeg)
                callback(if (result.isNotBlank()) result else "No text found in image.")
            } catch (e: Exception) {
                errorCallback(e.message ?: "OCR failed.")
            }
        }
    }
}
